File: codecheck-case/java/BKCT_ForwardNull/BKCT_ForwardNull_01.java
package java.BKCT_ForwardNull;

public class BKCT_ForwardNull_01 {

    public void badMethod1() {
        String str = null;
        if (str == null) {
            // bad
            System.out.println(str.length()); // 在为空的分支中使用
        }
    }

    public void badMethod2() {
        String str = null;
        if (str != null) {
            return;
        }
        // bad
        System.out.println(str.length()); // 在为空的分支中使用
    }

    public void badMethod3(String str) {
        if (str != null) {
            str = null;
        }
        // bad
        System.out.println(str.length()); // 在不再次检查是否为空的情况下使用
    }

    public void badMethod4(String str) {
        if (str == null) {

        }
        // bad
        System.out.println(str.length()); // 在不再次检查是否为空的情况下使用
    }

    public void goodMethod1() {
        String str = null;
        if (str != null) {
            // good
            System.out.println(str.length());
        }
    }

    public void goodMethod2() {
        String str = null;
        if (str == null) {
            return;
        }
        // good
        System.out.println(str.length());
    }

    public void goodMethod3(String str) {
        if (str == null) {
            str = "aaa";
        }
        // good
        System.out.println(str.length());
    }

    public void goodMethod4(String str) {
        if (str == null) {
            return;
        }
        // good
        System.out.println(str.length());
    }
}

File: codecheck-case/go/BKCT_ForwardNull/BKCT_ForwardNull_01.go
package BKCT_ForwardNull

import (
	"fmt"
)

func badMethod1() {
	var str *string
	if str == nil {
		// bad
		fmt.Println(len(*str)) // 在为空的分支中使用
	}
}

func badMethod2() {
	var str *string
	if str != nil {
		return
	}
	// bad
	fmt.Println(len(*str)) // 在为空的分支中使用
}

func badMethod3(str *string) {
	if str != nil {
		str = nil
	}
	// bad
	fmt.Println(len(*str)) // 在不再次检查是否为空的情况下使用
}

func badMethod4(str *string) {
	if str == nil {
		// do nothing
	}
	// bad
	fmt.Println(len(*str)) // 在不再次检查是否为空的情况下使用
}

func goodMethod1() {
	var str *string
	if str != nil {
		// good
		fmt.Println(len(*str))
	}
}

func goodMethod2() {
	var str *string
	if str == nil {
		return
	}
	// good
	fmt.Println(len(*str))
}

func goodMethod3(str *string) {
	if str == nil {
		temp := "aaa"
		str = &temp
	}
	// good
	fmt.Println(len(*str))
}

func goodMethod4(str *string) {
	if str == nil {
		return
	}
	// good
	fmt.Println(len(*str))
}
